horizomer/utils/tree.py

```diff
@@ -91,6 +144,66 @@ def intersect_trees(tree1, tree2):
     return (tree1_lap, tree2_lap)
 
 
+def collapse_short_branches(tree, cutoff):
+    """Collapse internal nodes with branch length below cutoff.
+
+    Parameters
+    ----------
+    tree : skbio.TreeNode
+        tree to search for nodes to collapse
+    cutoff : int or float
+        branch length cutoff under which node should be collapsed
+
+    Returns
+    -------
+    skbio.TreeNode
+        resulting tree with short branches collapsed
+
+    Notes
+    -------
+    Nodes without branch length are considered as having zero branch length and
+    will be collapsed.
+    """
+    tcopy = tree.copy()
+    nodes_to_collapse = []
+    for node in tcopy.non_tips():
+        if (node.length or 0.0) < cutoff:
+            nodes_to_collapse.append(node)
+    for node in nodes_to_collapse:
+        collapse(node)
+    return tcopy
+
+
+def collapse_low_support_nodes(tree, cutoff):
```

Comment:
very much reads like a code duplication which you might want to avoid by generalizing both collapse functions?

Reply:
That's why I have the `collapse` function in the beginning. But maybe you are saying that these two functions can be further simplified? Such as making another function `collapse_nodes` which takes a lambda function as parameter and takes care of the entire tree?

Reply:
The two functions have slightly different logics. When collapsing short branches, none branch lengths are collapsed too. But when collapsing low support nodes, non support (no label) are kept.

Moreover, I guess that if we remove those two functions, but let user freely combine criteria (e.g., collapsing short branches and low-support nodes and node labels reading "x" so on so forth) there may be logical problems. Although I cannot be very certain of it.

Reply:
One possibility is to have a function defined as follows

```
def collapse_by_value(tree, attr, cutoff):
""" 
    Parameters
    ------------
    tree : skbio.TreeNode
       Tree
    attr : str
       Attribute name
    cutoff : float
       cutoff value
    """
    value = getattr(tree, attr) # can be either support or length
```
However, this would force you to redecorate all of the support values directly on the trees.  Basically
```
for n in tree.postorder():
   n.support = float(n.name.split(':')[0])
```

Not entirely sure how this fits into the big picture.  But this is one way to reduce code-redundancy.

Reply:
I think the function suggested by Jamie sounds quite reasonable. Moreover, I think adding `support` as a property to the `TreeNode` class is a good idea. Even though my codes are standalone functions, they can be modified slightly to fit into the framework of scikit-bio. Plus, I don't think this will force re-decoration of all nodes. In the collapse functions I wrote, no existing node labels will be modified, including support values.

Reply:
I am using attributes extensively. You should be able to do:
`tree.support = 0.123`
or
`print(tree.support)`
with output 0.123
Be aware that tree.support might be `None`

Reply:
Can you show me how to add a property to the TreeNode data type without modifying scikit-bio itself? I wish to learn that.

Reply:
assume all tree nodes are decorated here with length and/or support (raise an exception otherwise) and check .attribute < cutoff where attribute is a parameter of the function. Thus, you would save code duplication and maybe recomputation of support values for later use